text/0000-build-observability.md

```diff
@@ -0,0 +1,374 @@
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes leveraging [OpenTelemetry](https://opentelemetry.io/) to
+grant platform operators and buildpack operators more insight into buildpack
+performance and behavior. This RFC describes new opt-in functionality
+for pack, the lifecycle, and the buildpack spec such that OpenTelemetry data may be
+exported to the build file system.
+
+# Definitions
+[definitions]: #definitions
+
+- [OpenTelemetry](https://opentelemetry.io/): A collection of APIs, SDKs, and tools that can be used it to instrument, generate, collect, and export telemetry data.
+- [Traces](https://opentelemetry.io/docs/concepts/signals/traces/): Telemetry
+  category that describes the path of software execution.
+
+
+# Motivation
+[motivation]: #motivation
+
+Buildpack authors and platform operators desire insight into usage, error
+scenarios, and performance of builds and buildpacks on their platform. The
+following questions are all important for these folks, but difficult to answer:
+
+- "Which buildpacks commonly fail to compile?"
+- "How often does a particular error scenario occur?"
+- "How long does each buildpack compile phase take?"
+- "How often is a certain buildpack used?"
+- "Which versions of Go are being installed?"
+- "How long does it take to download node_modules?"
+
+Instrumenting lifecycle and buildpacks with opt-in OpenTelemetry tracing will
+allow  platform operators to better understand performance and behavior of their
+builds and buildpacks and as a result, provide better service and build
+experiences.
+
+To protect privacy and prevent unnecessary collection of data, this
+functionality shall be optional and anonymous.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC aims to provide a solution for two types of OpenTelemetry traces:
+
+1) Lifecycle tracing: Buildpack-agnostic trace data like which buildpacks were
+available, which buildpacks were detected, how long the detect, build, or
+export phase took, and so on. This telemetry data may be exported by lifecycle.
+2) Buildpack tracing: Telemetry data specific to a buildpack like how long it
+took to download a language binary, which language version was selected, and so
+on. This telemetry data may be exported by buildpacks.
+
+Though the sources and contents of the telemetry data differ, both types may
+be emitted to the build file system in OpenTelemetry's [File Exporter
+Format](https://opentelemetry.io/docs/specs/otel/protocol/file-exporter/).
+
+In this solution, each lifecycle phase would write a `.jsonl` file with
+tracing data for that phase. For example, `lifecycle detector --telemetry`
+would write to `/layers/lifecycle-detect.jsonl`. Additionally each
+buildpack may also write tracing data to it's own `.jsonl` files (at
+`/layers/#{buildpack-id}@#{buildpack-version}.jsonl`).
```

Comment:
Should this include the phase as well? E.g., `/layers/#{buildpack-id}@#{buildpack-version}-detect.jsonl`. And what about extensions?

Reply:
Bikeshedding:

```
<layers>/
└── tracing
    ├── buildpacks
    │   ├── other-id@other-version-detect.jsonl.   # this buildpack didn't detect
    │   ├── some-id@some-version-build.jsonl
    │   └── some-id@some-version-detect.jsonl
    └── extensions
        ├── some-id@some-version-detect.jsonl    # extensions and buildpacks can share the same ID
        └── some-id@some-version-generate.jsonl
```

This does add one more reserved buildpack ID, i.e., forbidden directory under `<layers>`. But, we already have `app`, `config`, `extended`, `generated`, and `sbom` ...is there harm in one more? (Really)

Eventually I would love to solve the reserved buildpack ID problem, maybe by moving at least `extended`, `generated`, and `sbom` under something like `<layers>/cnb`. But I don't think we need to do it right here and right now.

Reply:
> Should this include the phase as well? E.g., /layers/#{buildpack-id}@#{buildpack-version}-detect.jsonl. And what about extensions?

I think this might be useful for a platform operator -- their OTel collector could possibly enrich traces with attributes like `buildpack-id`, `buildpack-version`, `buildpack-phase`, `buildpack_or_extension`  based on the filename. Though, they would have to build that in to their collector. And in my mind, it's fairly likely that each buildpack would already include attributes like these in the tracing data they emit anyway.

That's about the only use-case I could see for this differentiation. So it's maybe not super useful in my mind, but it's low complexity for buildpack authors and telemetry consumers. It's probably worth the tradeoff.

Reply:
> This does add one more reserved buildpack ID, i.e., forbidden directory under <layers>.

I like the `/<layers>/tracing` idea. I can add that. 

Since there are a bunch of `lifecycle-*.jsonl` files, might as well put them in a directory too?

```
<layers>/
└── tracing
    ├── buildpacks
    │   ├── other-id@other-version-detect.jsonl
    │   ├── some-id@some-version-build.jsonl
    │   └── some-id@some-version-detect.jsonl
    ├── extensions
    │   ├── some-id@some-version-detect.jsonl
    │   └── some-id@some-version-generate.jsonl
    └── lifecycle
        ├── analyze.jsonl
        ├── build.jsonl
        ├── detect.jsonl
        ├── export.jsonl
        ├── extend.jsonl
        └── restore.jsonl
```

Reply:
Sounds good to me :)